```
The failing job is due to a linker error where the symbol mooncake::Environ::Get() is undefined.

Here are the specific lines from the logs pointing to the issue:

Text
/usr/bin/ld: /home/runner/work/Mooncake/Mooncake/build/mooncake-transfer-engine/src/libtransfer_engine.a(rdma_context.cpp.o): in function `mooncake::RdmaContext::registerMemoryRegionInternal(void*, unsigned long, int, mooncake::MemoryRegionMeta&)':
/home/runner/work/Mooncake/Mooncake/mooncake-transfer-engine/src/transport/rdma_transport/rdma_context.cpp:248: undefined reference to `mooncake::Environ::Get()'
...
/home/runner/work/Mooncake/Mooncake/mooncake-transfer-engine/src/transport/rdma_transport/rdma_context.cpp:663: undefined reference to `mooncake::Environ::Get()'
collect2: error: ld returned 1 exit status
Solution
This error occurs when the code compiles successfully (so the header is found), but the compiler cannot find the compiled implementation of the function during the linking phase.

Here's how to fix it:

Check the CMakeLists.txt: Ensure that the C++ file defining mooncake::Environ::Get() (e.g., environ.cpp) is actually included in the source file list for building libtransfer_engine.a (or whichever library/target builds rdma_context.cpp).
Missing Dependency Linkage: If mooncake::Environ::Get() is defined in another library within your project, make sure target_link_libraries includes that library when linking the final executable or library target that complains.
You will need to update the relevant CMakeLists.txt to either append the missing .cpp implementation file or link the missing internal library containing Environ.
```
@copilot+claude-sonnet-4.6 Solve it.

--- a/mooncake-p2p-store/build.sh
+++ b/mooncake-p2p-store/build.sh
@@ -34,7 +34,7 @@ fi
 EXT_LDFLAGS="-L$BUILD_DIR/mooncake-transfer-engine/src"
 EXT_LDFLAGS+=" -L$BUILD_DIR/mooncake-transfer-engine/src/common/base"
 EXT_LDFLAGS+=" -L$BUILD_DIR/mooncake-common"
-EXT_LDFLAGS+=" -ltransfer_engine -lbase -lasio -lstdc++ -lnuma -lglog -libverbs -ljsoncpp"
+EXT_LDFLAGS+=" -ltransfer_engine -lbase -lasio -lstdc++ -lnuma -lglog -libverbs -ljsoncpp -lmooncake_common"
 
 if [ -d "/usr/local/cuda/lib64/stubs" ]; then
     EXT_LDFLAGS+=" -L/usr/local/cuda/lib64/stubs"

--- a/mooncake-store/go/build.sh
+++ b/mooncake-store/go/build.sh
@@ -40,7 +40,7 @@ CGO_LDFLAGS+=" -L${BUILD_DIR}/mooncake-store/src/cachelib_memory_allocator"
 CGO_LDFLAGS+=" -L${BUILD_DIR}/mooncake-transfer-engine/src"
 CGO_LDFLAGS+=" -L${BUILD_DIR}/mooncake-transfer-engine/src/common/base"
 CGO_LDFLAGS+=" -L${BUILD_DIR}/mooncake-common"
-CGO_LDFLAGS+=" -lmooncake_store -lcachelib_memory_allocator -ltransfer_engine -lbase -lasio"
+CGO_LDFLAGS+=" -lmooncake_store -lcachelib_memory_allocator -ltransfer_engine -lbase -lasio -lmooncake_common"
 CGO_LDFLAGS+=" -lstdc++ -lnuma -lglog -lgflags -libverbs -ljsoncpp -lzstd -lcurl"
 
 if [ -d "/usr/local/cuda/lib64" ]; then

--- a/mooncake-transfer-engine/CMakeLists.txt
+++ b/mooncake-transfer-engine/CMakeLists.txt
@@ -63,7 +63,7 @@ endif()
 
 # Non-peermem requires cuMem allocated memory pointers; since WITH_NVIDIA_PEERMEM
 # is now a runtime environment variable, always build when GPU support is present.
-if (USE_MNNVL OR USE_CUDA OR USE_HIP OR USE_MUSA)
+if (USE_MNNVL OR USE_CUDA OR USE_HIP OR USE_MUSA OR USE_MACA)
   add_subdirectory(nvlink-allocator)
 endif()
 

--- a/mooncake-transfer-engine/nvlink-allocator/CMakeLists.txt
+++ b/mooncake-transfer-engine/nvlink-allocator/CMakeLists.txt
@@ -5,12 +5,15 @@ if(USE_HIP)
   list(APPEND _extra_build_opts --use-hipcc)
 elseif(USE_MUSA)
   list(APPEND _extra_build_opts --use-mcc)
+elseif(USE_MACA)
+  list(APPEND _extra_build_opts --use-maca)
 endif()
 
 set(_enable_nvlink_allocator_build FALSE)
 if(USE_CUDA
    OR USE_HIP
-   OR USE_MUSA)
+   OR USE_MUSA
+   OR USE_MACA)
   set(_enable_nvlink_allocator_build TRUE)
 endif()
 

--- a/mooncake-transfer-engine/nvlink-allocator/build.sh
+++ b/mooncake-transfer-engine/nvlink-allocator/build.sh
@@ -8,6 +8,7 @@ source "$(dirname "$(readlink -f "$0")")/../scripts/allocator_build_common.sh"
 USE_NVCC=false
 USE_HIPCC=false
 USE_MCC=false
+USE_MACA=false
 CI_BUILD=false
 
 if [[ "$1" == "--use-nvcc" ]]; then
@@ -18,6 +19,9 @@ elif [[ "$1" == "--use-hipcc" ]]; then
     shift
 elif [[ "$1" == "--use-mcc" ]]; then
     USE_MCC=true
+    shift
+elif [[ "$1" == "--use-maca" ]]; then
+    USE_MACA=true
     shift
 elif [[ "$1" == "--ci-build" ]]; then
     CI_BUILD=true
@@ -45,6 +49,11 @@ elif [ "$USE_HIPCC" = true ]; then
     hipcc "$OUTPUT_DIR/nvlink_allocator.cpp" -o "$OUTPUT_DIR/nvlink_allocator.so" -shared -fPIC -lamdhip64 -I/opt/rocm/include ${INCLUDE_FLAGS} -DUSE_HIP=1
 elif [ "$USE_MCC" = true ]; then
     mcc "$CPP_FILE" -o "$OUTPUT_DIR/nvlink_allocator.so" --shared -fPIC -lmusa -I/usr/local/musa/include ${INCLUDE_FLAGS} -DUSE_MUSA=1
+elif [ "$USE_MACA" = true ]; then
+    MACA_ROOT=${MACA_HOME:-/opt/maca}
+    g++ "$CPP_FILE" -o "$OUTPUT_DIR/nvlink_allocator.so" --shared -fPIC \
+        -I"${MACA_ROOT}/include" ${INCLUDE_FLAGS} \
+        -L"${MACA_ROOT}/lib64" -L"${MACA_ROOT}/lib" -lmcruntime -DUSE_MACA=1
 else
     # Default g++ build
     g++ "$CPP_FILE" -o "$OUTPUT_DIR/nvlink_allocator.so" --shared -fPIC -lcuda -I/usr/local/cuda/include ${INCLUDE_FLAGS} -DUSE_CUDA=1

--- a/mooncake-transfer-engine/src/CMakeLists.txt
+++ b/mooncake-transfer-engine/src/CMakeLists.txt
@@ -48,6 +48,7 @@ target_link_libraries(
   PUBLIC base
          transport
          rdma_transport
+         mooncake_common
          ibverbs
          glog::glog
          gflags::gflags

